src/charm.py

```diff
@@ -1,115 +1,117 @@
 
-"""Charm the service.
-
-Refer to the following post for a quick-start guide that will help you
-develop a new k8s charm using the Operator Framework:
-
-https://discourse.charmhub.io/t/4208
-"""
 
 import logging
+import typing
 
 import ops
+from ops.main import main
 
-# Log messages can be retrieved using juju debug-log
-logger = logging.getLogger(__name__)
+import agent_observer
+import service
+from charm_state import (
+    AGENT_RELATION,
+    InvalidStateError,
+    State,
+    get_agent_interface_dict_from_metadata,
+)
 
-VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]
+logger = logging.getLogger()
 
 
-class IsCharmsTemplateCharm(ops.CharmBase):
-    """Charm the service."""
+class JenkinsAgentCharm(ops.CharmBase):
+    """Charm Jenkins agent."""
 
-    def __init__(self, *args):
-        """Construct.
+    def __init__(self, *args: typing.Any):
+        """Initialize the charm and register event handlers.
 
         Args:
-            args: Arguments passed to the CharmBase parent constructor.
+            args: Arguments to initialize the charm base.
         """
         super().__init__(*args)
-        self.framework.observe(self.on.httpbin_pebble_ready, self._on_httpbin_pebble_ready)
+        try:
+            self.state = State.from_charm(self)
+        except InvalidStateError as e:
+            logger.debug("Error parsing charm_state %s", e)
+            self.unit.status = ops.BlockedStatus(e.msg)
+            return
+
+        self.jenkins_agent_service = service.JenkinsAgentService(self.state)
+        self.agent_observer = agent_observer.Observer(self, self.state, self.jenkins_agent_service)
+
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.start, self._on_start)
         self.framework.observe(self.on.config_changed, self._on_config_changed)
-
-    def _on_httpbin_pebble_ready(self, event: ops.PebbleReadyEvent):
-        """Define and start a workload using the Pebble API.
-
-        Change this example to suit your needs. You'll need to specify the right entrypoint and
-        environment configuration for your specific workload.
-
-        Learn more about interacting with Pebble at at https://juju.is/docs/sdk/pebble.
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
+        self.framework.observe(self.on.update_status, self._on_update_status)
+
+    def _on_install(self, _: ops.InstallEvent) -> None:
+        """Handle install event, setup the agent service."""
+        try:
+            self.jenkins_agent_service.install()
+        except service.PackageInstallError as e:
+            logger.debug("Error installing the agent service %s", e)
+            self.unit.status = ops.ErrorStatus("Error installing the agent service")
+
+    def _on_config_changed(self, _: ops.ConfigChangedEvent) -> None:
+        """Handle config changed event. Update the agent's label in the relation's databag."""
```

Comment:
Docstring is not complete. Isn't CI pointing this out? If not, something might be misconfigured